### PR TITLE
Create a backend for the pr_curves plugin

### PR DESCRIPTION
**Pull request by chihuahua**
Made PrCurvesPlugin, which serves data for the PR Curves dashboard. Made a metadata.py file that centralizes information such as the plugin name and the indices used to obtain precision and recall from tensor data.

This PR is a part of dividing https://github.com/tensorflow/tensorboard/pull/334 (which contains resourceful input from @wchargin) into smaller PRs.

### CODE REVIEW

File: tensorboard/plugins/pr_curve/metadata.py [hunk 1]
```diff
@@ -0,0 +1,55 @@
...
+    display_name: The display name used in TensorBoard.
+    description: The description to show in TensorBoard.
+    num_thresholds: The number of thresholds to use for PR curves.
+
+  Returns:
+    A `tf.SummaryMetadata` protobuf object.
+  """
+  pr_curve_plugin_data = plugin_data_pb2.PrCurvePluginData(
+      version=PROTO_VERSION, num_thresholds=num_thresholds)
+  content = json_format.MessageToJson(pr_curve_plugin_data)
```
Comment (wchargin) [inline tensorboard/plugins/pr_curve/metadata.py]: This should be `content = pr_curve_plugin_data.SerializeToString()`, for consistency with our stated best-practice and consistency with all other plugins.

File: tensorboard/plugins/pr_curve/pr_curves_plugin.py [hunk 2]
```diff
@@ -0,0 +1,198 @@
...
+    tag = request.args.get('tag')
+    if not tag:
+      return http_util.Respond(
+          request, 'No tag provided when fetching PR curve data', 400)
+
+    try:
+      response = http_util.Respond(
+          request, self.pr_curves_impl(runs, tag), 'application/json')
+    except ValueError as e:
+      return http_util.Respond(request, '%s' % e, 'text/plain', 400)
```
Comment (wchargin) [inline tensorboard/plugins/pr_curve/pr_curves_plugin.py]: `'%s' % e` is more easily written as `str(e)`.

File: tensorboard/plugins/pr_curve/pr_curves_plugin.py [hunk 3]
```diff
@@ -0,0 +1,198 @@
...
+        request, self.tags_impl(), 'application/json')
+
+  def tags_impl(self):
+    """Creates the JSON object for the tags route response.
+
+    Returns:
+      The JSON object for the tags route response.
+    """
+    all_runs = self._multiplexer.PluginRunToTagToContent(
+        PrCurvesPlugin.plugin_name)
```
Comment (wchargin) [inline tensorboard/plugins/pr_curve/pr_curves_plugin.py]: This should be `metadata.PLUGIN_NAME`. I know that they have the same value, but the argument to `PluginRunToTagToContent` corresponds to the string stored in the `plugin_data` of the summaries, not the route used to fetch data. If you changed `PrCurvesPlugin.plugin_name` to be something else, the value of this argument should still be the original `metadata.PLUGIN_NAME`.

Likewise elsewhere in this file.

File: tensorboard/plugins/pr_curve/pr_curves_plugin.py [hunk 4]
```diff
@@ -0,0 +1,198 @@
...
+        # This run lacks data for this plugin.
+        continue
+
+      # Just use the list of tensor events for any of the tags to determine the
+      # steps to list for the run. The steps are often the same across tags for
+      # each run, albeit the user may elect to sample certain tags differently
+      # within the same run. If the latter occurs, TensorBoard will show the
+      # actual step of each tag atop the card for the tag.
+      tensor_events = self._multiplexer.Tensors(
+          run, list(tag_to_content.keys())[0])
```
Comment (wchargin) [inline tensorboard/plugins/pr_curve/pr_curves_plugin.py]: This is nondeterministic, which rubs me the wrong way. Consider: `min(six.iterkeys(tag_to_content))`.

(General note: If you were to not apply this comment, prefer `next(six.iterkeys(tag_to_content))` to what's written here. No need to serialize to list, duplicate the list (in Python 2), and then grab the first element.)

File: tensorboard/plugins/pr_curve/pr_curves_plugin_test.py [hunk 5]
```diff
@@ -0,0 +1,187 @@
...
+from tensorboard.plugins.pr_curve import pr_curve_demo
+from tensorboard.plugins.pr_curve import pr_curves_plugin
+
+
+class PrCurvesPluginTest(tf.test.TestCase):
+
+  def setUp(self):
+    super(PrCurvesPluginTest, self).setUp()
+    logdir = os.path.join(self.get_temp_dir(), 'logdir')
+    tf.reset_default_graph()
```
Comment (wchargin) [inline tensorboard/plugins/pr_curve/pr_curves_plugin_test.py]: This is not needed; it's done in the superclass setup.

File: tensorboard/plugins/pr_curve/pr_curves_plugin_test.py [hunk 6]
```diff
@@ -0,0 +1,187 @@
...
+
+    Args:
+      expected_step: The expected step.
+      expected_precision: A list of float precision values.
+      expected_recall: A list of float recall values.
+      pr_curve_entry: The PR curve entry to evaluate.
+    """
+    self.assertEqual(expected_step, pr_curve_entry['step'])
+    # We use an absolute error instead of a relative one because the expected
+    # values are small. The default relative error (trol) of 1e-7 yields many
```
Comment (wchargin) [inline tensorboard/plugins/pr_curve/pr_curves_plugin_test.py]: s/trol/rtol?

File: tensorboard/plugins/pr_curve/pr_curves_plugin.py [hunk 7]
```diff
@@ -0,0 +1,198 @@
+# Copyright 2017 Google Inc. All Rights Reserved.
```
Comment (wchargin) [inline tensorboard/plugins/pr_curve/pr_curves_plugin.py]: Could you add an `http_api.md`? (Copy-paste and modify one of the other plugins'.)

File: tensorboard/plugins/pr_curve/pr_curves_plugin_test.py [hunk 8]
```diff
@@ -0,0 +1,187 @@
...
+        expected_recall=[1.0, 0.8133333, 0.2133333, 0.0266666, 0.0],
+        pr_curve_entry=entries[2])
+
+  def testPrCurvesRaisesValueErrorWhenNoData(self):
+    """Tests that the method for obtaining PR curve data raises a ValueError.
+
+    The handler should raise a ValueError when no PR curve data can be found
+    for a certain run-tag combination.
+    """
+    with self.assertRaises(ValueError):
```
Comment (wchargin) [inline tensorboard/plugins/pr_curve/pr_curves_plugin_test.py]: You might prefer `with six.assertRaisesRegex(self, ValueError, r'No PR curves could be fetched')` to ensure that the error is the one that you're looking for.

File: tensorboard/plugins/pr_curve/summary.py [hunk 9]
```diff
@@ -58,7 +57,7 @@ def op(
     predictions: A float32 `Tensor` whose values are in the range `[0, 1]`.
         Dimensions must match those of `labels`.
     num_thresholds: Number of thresholds, evenly distributed in `[0, 1]`, to
-        compute PR metrics for. Should be `>= 2`. This value should be a 
+        compute PR metrics for. Should be `>= 2`. This value should be a
```
Comment (wchargin) [inline tensorboard/plugins/pr_curve/summary.py]: I liked the idea of handling these in #389. Perhaps you have a rogue commit in this PR?

Ideally, the "add backend" commit should make no changes to `summary.py`, and the "add frontend" commit (if separate) should make no changes to `pr_curves_plugin.py`/etc.

File: tensorboard/plugins/pr_curve/metadata.py (hunk 1 above)
Comment (chihuahua) [inline tensorboard/plugins/pr_curve/metadata.py]: Done.

File: tensorboard/plugins/pr_curve/summary.py (hunk 9 above)
Comment (chihuahua) [inline tensorboard/plugins/pr_curve/summary.py]: Indeed, I am going to let #389 handle these style changes.

File: tensorboard/plugins/pr_curve/pr_curves_plugin.py [hunk 10]
```diff
@@ -0,0 +1,195 @@
...
+    return http_util.Respond(
+        request, self.tags_impl(), 'application/json')
+
+  def tags_impl(self):
+    """Creates the JSON object for the tags route response.
+
+    Returns:
+      The JSON object for the tags route response.
+    """
+    all_runs = self._multiplexer.PluginRunToTagToContent(metadata.PLUGIN_NAME)
```
Comment (chihuahua) [inline tensorboard/plugins/pr_curve/pr_curves_plugin.py]: Done. Thank you for noting that! I also updated http.md.

File: tensorboard/plugins/pr_curve/http_api.md [hunk 11]
```diff
@@ -0,0 +1,97 @@
...
+## `/data/plugin/scalars/pr_curves`
+
+Retrieves a JSON object mapping run name to a list of PR curve data entries (one
+entry per step). This route requires a `tag` GET parameter as well as at least
+one `run` GET parameter to specify both the tag and list of runs to retrieve
+data for. 
+
+Each PR data entry contains the following properties.
+
+* **wall_time**: The wall time (number) in seconds since the epoch at which data
```
Comment (chihuahua) [inline tensorboard/plugins/pr_curve/http_api.md]: Done. A `<dl>` would seem apt if possible.

File: tensorboard/plugins/pr_curve/pr_curves_plugin.py [hunk 12]
```diff
@@ -0,0 +1,225 @@
...
+        # This run lacks data for this plugin.
+        continue
+
+      # Just use the list of tensor events for any of the tags to determine the
+      # steps to list for the run. The steps are often the same across tags for
+      # each run, albeit the user may elect to sample certain tags differently
+      # within the same run. If the latter occurs, TensorBoard will show the
+      # actual step of each tag atop the card for the tag.
+      tensor_events = self._multiplexer.Tensors(
+          run, list(tag_to_content.keys())[0])
```
Comment (wchargin) [inline tensorboard/plugins/pr_curve/pr_curves_plugin.py]: This was reverted here: https://github.com/tensorflow/tensorboard/commit/382b13d06c206de4b82d42adec5f026652e38f0b#diff-9b643340c1c8537af733696bb8c0c641R166

I'm pausing review on the suspicion that this is a rogue commit (at a quick scan, its contents don't seem to match its title). Let me know when the diff is correct.

Comment (chihuahua) [inline tensorboard/plugins/pr_curve/pr_curves_plugin.py]: Thank you for noticing! I undid the rogue changes in a recent commit. What happened was that I had copied some code from the PR with the demo, which did not include some of the recent changes in this particular PR. I need to be more cognizant of avoiding rogue changes and commits.

File: tensorboard/plugins/pr_curve/http_api.md [hunk 13]
```diff
@@ -0,0 +1,124 @@
...
+
+## `/data/plugin/pr_curves/tags`
+
+Retrieves a JSON object whose keys are the names of all the runs (regardless of
+whether they have available PR curve data). The values of that object are also
+objects whose keys are tag strings (associated with the run) and whose values
+are objects with 2 keys: `displayName` and `description` (associated with the
+run-tag combination).
+
+The `displayName` is shown atop individual plots in TensorBoard. The description
```
Comment (chihuahua) [inline tensorboard/plugins/pr_curve/http_api.md]: Done.

File: tensorboard/plugins/pr_curve/pr_curves_plugin.py [hunk 14]
```diff
@@ -0,0 +1,225 @@
...
+    
+    Arguments:
+      tensor_event: The tensor event for the time entry.
+      initial_wall_time: The wall time of the first event in seconds since the
+        epoch. Used to compute relative time.
+    """
+    return {
+      'step': tensor_event.step,
+      'wall_time': tensor_event.wall_time,
+      'relative': tensor_event.wall_time - initial_wall_time,
```
Comment (wchargin) [inline tensorboard/plugins/pr_curve/pr_curves_plugin.py]: Why compute the relative times on the server? No other plugin does this, as far as I know; we use `ChartHelpers.relativeAccessor` to do this on the frontend: https://github.com/tensorflow/tensorboard/blob/542ac06be716e8b28114bacfcb02535a608c1e12/tensorboard/plugins/scalar/vz_line_chart/vz-chart-helpers.ts#L155-L169

Comment (chihuahua) [reply tensorboard/plugins/pr_curve/pr_curves_plugin.py]: That is true. The recent commit does away with computing relative on the backend. 

I think on the frontend, we could have pr-curve-card logic compute relative values. We could also generalize the logic within vz-chart-helpers, although the relative time in the PR curve UI changes the display in the sliders and doesn't modify the UI of the vz line chart.

**Pull request merged**